### PR TITLE
Add fine-tuning option

### LINKED ISSUE #38: add fine tuning

### PR DESCRIPTION
**Pull request by ElliottKasoar**
Aims to resolve #38

- Adds `--fine-tune` option to `janus train`
  - Given the current differences, this feels more natural than an entirely different command
  - Currently only checks that `foundation_model` is in the config for fine-tuning, and not present otherwise, and that the value of `foundation_model` is either a valid MACE-MP/MACE-OFF string or a path that exists
  - We could load the model (if given a path, or even if given a valid string), but we would be unable to pass this directly to MACE

- Adds tests based on data and configs from @alinelena in #38
  - Modifications include:
    - File names and locations for train/test/validation data
    -  `max_num_epochs` reduced from 10 to 1 for both configs
    - Uses `mace_mp_small.model` model for fine-tuning, as we use it elsewhere
    - Changes `default_dtype` to float64 to match model for fine-tuning
  - Skipped if unable to load run_train from MACE, so not currently tested in CI, but can be tested localled
  - The fine-tuning test is quite slow (~1 minute) - are there ways to reduce this without inhibiting the function of the test? It may be inevitably given the task

### CODE REVIEW

File: tests/test_train_cli.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+def test_train():
+    """Test MLIP training."""
+    model = "test.model"
+    compiled_model = "test_compiled.model"
+
+    assert not Path(model).exists()
+    assert not Path(compiled_model).exists()
+    assert not Path("logs").exists()
+    assert not Path("results").exists()
+    assert not Path("checkpoints").exists()
```
Comment (oerc0122) [inline tests/test_train_cli.py:41]: Prefer making variables of these paths, particularly since they're reused later

File: tests/test_train_cli.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+from pathlib import Path
+import shutil
+
+import pytest
+from typer.testing import CliRunner
+
+try:
+    from mace.cli.run_train import run as run_train  # pylint: disable=unused-import
+
+    SKIP_TESTS = False
```
Comment (oerc0122) [inline tests/test_train_cli.py:12]: More informative name wanted here.

**Review by oerc0122 — COMMENTED** [review]
Looks fine, just a couple of stylistic things.

File: tests/data/mlip_fine_tune.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+clip_grad: 100
+error_table: PerAtomRMSE
+lr: 0.005
+scaling: rms_forces_scaling
+batch_size: 4
+max_num_epochs: 1
+ema: True
+ema_decay: 0.995
+amsgrad: True
+default_dtype: float64
```
Comment (alinelena) [inline tests/data/mlip_fine_tune.yml:27]: make float32
```suggestion
default_dtype: float32
```

Comment (ElliottKasoar) [reply tests/data/mlip_fine_tune.yml:27]: I get a type mismatch:

```
  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/mace/cli/run_train.py", line 686, in run
    tools.train(

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/mace/tools/train.py", line 148, in train
    valid_loss, eval_metrics = evaluate(
                               ^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/mace/tools/train.py", line 357, in evaluate
    output = model(
             ^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/mace/modules/models.py", line 340, in forward
    node_feats = self.node_embedding(data["node_attrs"])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/mace/modules/blocks.py", line 44, in forward
    return self.linear(node_attrs)
           ^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/e3nn/o3/_linear.py", line 280, in forward
    return self._compiled_main(features, weight, bias)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "<eval_with_key>.26", line 15, in forward
    reshape_3 = reshape_1.reshape((4, 128));  reshape_1 = None
    mul = reshape_2 * 0.5;  reshape_2 = None
    tensordot = torch.functional.tensordot(mul, reshape_3, ([1], [0]), out = None);  mul = reshape_3 = None
                ~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    permute = tensordot.permute([0, 2, 1]);  tensordot = None
    reshape_4 = permute.reshape(getitem_1, 128);  permute = getitem_1 = None
  File "/home/ek/PSDI/venv_janus_11_new/lib/python3.11/site-packages/torch/functional.py", line 1213, in tensordot

    if out is None:
        return _VF.tensordot(a, b, dims_a, dims_b)  # type: ignore[attr-defined]
               ~~~~~~~~~~~~~ <--- HERE
    else:
        return _VF.tensordot(a, b, dims_a, dims_b, out=out)  # type: ignore[attr-defined]
RuntimeError: both inputs should have same dtype
```

Comment (alinelena) [reply tests/data/mlip_fine_tune.yml:27]: ok leave it to 64 I suspect the issue is of mixing model precision

File: tests/test_train_cli.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+
+@pytest.mark.skipif(MACE_IMPORT_ERROR, reason="Requires updated version of MACE")
+def test_fine_tune():
+    """Test MLIP fine-tuning."""
+    model = "test-finetuned.model"
+    compiled_model = "test-finetuned_compiled.model"
+    logs_path = "logs"
+    results_path = "results"
+    checkpoints_path = "checkpoints"
```
Comment (oerc0122) [inline tests/test_train_cli.py:99]: Been playing around in other repos with [`tempfile.TemporaryDirectory`](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) which may be what we want with these rather than the `try`-`finally`s. Probably better to do globally as another PR though.

**Review by oerc0122 — APPROVED** [review]

File: tests/test_train_cli.py (hunk 4 above)
Comment (ElliottKasoar) [reply tests/test_train_cli.py:99]: In this case, MACE is creating these directories, so even if they're options that we can set, while we're reliant on passing a predefined config file, this wouldn't be particularly straightforward to make use.

Worth looking into though, I agree.

**Review by oerc0122 — APPROVED** [review]
